straxen/mongo_storage.py

```diff
@@ -0,0 +1,358 @@
+
+
+@export
+class Uploader(GridFsInterface):
+    """
+    Class to upload files to GridFs
+    """
+
+    def __init__(self, readonly=False, *args, **kwargs):
+        # Same as parent. Just check the redonly_argument
+        if readonly:
+            raise PermissionError(
+                "How can you upload if you want to operate in readonly?")
+        super().__init__(*args, readonly=readonly, **kwargs)
+
+    def upload_from_dict(self, file_path_dict):
+        """
+        Upload all files in the dictionary to the database.
+        :param file_path_dict: dict, dictionary of paths to upload. The 
+            dict should be of the format:
+            file_path_dict = {'config_name':  '/the_config_path', ...}
+        :return: None
+        """
+        if not isinstance(file_path_dict, dict):
+            raise ValueError(f'file_path_dict must be dict of form '
+                             f'"dict(NAME=ABSOLUTE_PATH,...)". Got '
+                             f'{type(file_path_dict)} instead')
+
+        for config, abs_path in tqdm(file_path_dict.items()):
+            # We need to do this expensive check here. It is not enough
+            # to just check that the file is stored under the 
+            # 'config_identifier'. What if the file changed? Then we 
+            # want to upload a new file! Otherwise we could have done 
+            # the self.config_exists-query. If it turns out we have the 
+            # exact same file, forget about uploading it.
+            if self.md5_stored(abs_path):
+                continue
+            else:
+                # This means we are going to upload the file because its
+                # not stored yet.
+                try:
+                    self.upload_single(config, abs_path)
+                except (CouldNotLoadError, ConfigTooLargeError):
+                    # Perhaps we should fail then?
+                    warn(f'Cannot upload {config}')
+
+    def upload_single(self, config, abs_path):
+        """
+        Upload a single file to gridfs
+        :param config: str, the name under which this file should be 
+            stored
+        :param abs_path: str, the absolute path of the file 
+        """
+        doc = self.document_format(config)
+        if not os.path.exists(abs_path):
+            raise CouldNotLoadError(f'{abs_path} does not exits')
+
+        print(f'uploading {config}')
+        with open(abs_path, 'rb') as file:
+            self.grid_fs.put(file, **doc)
+
+
+@export
+class Downloader(GridFsInterface):
+    """
+    Class to download files from GridFs
+    """
+
+    def __init__(self,
+                 store_files_at=None,
+                 *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # We are going to set a place where to store the files. It's 
+        # either specified by the user or we use these defaults:
+        if store_files_at is None:
+            store_files_at = ('./resource_cache',

```

Comment:
Would you like to give the user the opportunity to extend this with another path?

Reply:
Sure, one for example might really want to specify that some files are stored elsewhere. Than this is easy to do at the init of the `Downloader`. I don't think it's a problem if we allow a user to cache files wherever the user desires.